src/cppnv/EnvReader.cc

```diff
@@ -0,0 +1,813 @@
+#include "EnvReader.h"
+
+
+namespace cppnv {
+EnvReader::read_result EnvReader::read_pair(EnvStream* file,
+                                            const EnvPair* pair) {
+  const read_result result = read_key(file, pair->key);
+  if (result == fail || result == empty) {
+    return fail;
+  }
+  if (result == comment_encountered) {
+    return comment_encountered;
+  }
+  if (result == end_of_stream_key) {
+    return end_of_stream_key;
+  }
+
+  if (result == end_of_stream_value) {
+    return success;
+  }
+  //  trim right side of key
+  while (pair->key->key_index > 0) {
+    if (pair->key->key->at(pair->key->key_index - 1) != ' ') {
+      break;
+    }
+    pair->key->key_index--;
+  }
+  if (!pair->key->has_own_buffer()) {
+    const auto tmp_str = new std::string(pair->key->key_index, '\0');
+
+    tmp_str->replace(0,
+                     pair->key->key_index,
+                     *pair->key->key,
+                     0,
+                     pair->key->key_index);
+
+    pair->key->set_own_buffer(tmp_str);;
+  } else {
+    pair->key->clip_own_buffer(pair->key->key_index);
+  }
+  pair->value->value->clear();
+  const read_result value_result = read_value(file, pair->value);
+  if (value_result == end_of_stream_value) {
+    return end_of_stream_value;
+  }
+  if (value_result == comment_encountered || value_result == success) {
+    if (!pair->value->has_own_buffer()) {
+      const auto tmp_str =
+          new std::string(pair->value->value_index, '\0');
+
+      tmp_str->replace(0,
+                       pair->value->value_index,
+                       *pair->value->value,
+                       0,
+                       pair->value->value_index);
+
+      pair->value->
+            set_own_buffer(tmp_str);
+    } else {
+      pair->value->clip_own_buffer(pair->value->value_index);
+    }
+    remove_unclosed_interpolation(pair->value);
+    return success;
+  }
+  if (value_result == empty) {
+    remove_unclosed_interpolation(pair->value);
+    return empty;
+  }
+  if (value_result == end_of_stream_key) {
+    remove_unclosed_interpolation(pair->value);
+    return end_of_stream_key;
+  }
+
+  remove_unclosed_interpolation(pair->value);
+  return fail;
+}
+
+
+int EnvReader::read_pairs(EnvStream* file, std::vector<EnvPair*>* pairs) {
+  int count = 0;
+  auto buffer = std::string(256, '\0');
+
```

Comment:
since we're clearing this buffer on loop, i think we can just use `std::string();`  rather than write empty string with that size :)

Reply:
I chose that constructor to avoid additional allocations. Clear doesn't shrink just sets the size to 0. If you believe I am mistaken I can change it. A std::string only holds 16 bytes and increases by 2x so in a loop like this it could allocate several times.